### PR TITLE
Fix data source rewind across more than one block

### LINKED ISSUE #4994: [Bug] Assertion about dynamic data sources failed
### Bug report

A subgraph deployed into my graph node is failing with

```
thread 'QmbrEG3Ph195yb341q1utKWu1WiupaeSeYRGRTdZaDASqC[2737]' panicked at 'assertion failed: self.hosts.last().and_then(|h| h.creation_block_number()) <=
    data_source.creation_block()', /graph-node/core/src/subgraph/context/instance.rs:136:9
```

I'm also seeing issues with the creation of runtime hosts.

### Relevant log output

```Shell
Nov 16 09:27:23.651 WARN no runtime host created, there is already a runtime host instantiated for this data source, address: 6261666b726569676875626d677572717070336b6d6f666c637663356635766c366661376b736f34657077737a366235366e73706d377669346c75, name: MetadataInfo, sgd: 2737, subgraph_id: QmbrEG3Ph195yb341q1utKWu1WiupaeSeYRGRTdZaDASqC, component: SubgraphInstanceManager
Nov 16 09:27:23.810 WARN no runtime host created, there is already a runtime host instantiated for this data source, address: 6261666b726569633671767073636c6d6f6570787473726e786d6f7a70326f3566777579756c6164753733773762677461366b35796c766c6c7779, name: MetadataInfo, sgd: 2737, subgraph_id: QmbrEG3Ph195yb341q1utKWu1WiupaeSeYRGRTdZaDASqC, component: SubgraphInstanceManager
Nov 16 09:27:31.856 WARN no runtime host created, there is already a runtime host instantiated for this data source, address: 6261666b726569676d32717264357366656c756a35367135783274673369706674676e6b7a6b37697a6b7a6e7868346a676f697965776d32697234, name: MetadataInfo, sgd: 2737, subgraph_id: QmbrEG3Ph195yb341q1utKWu1WiupaeSeYRGRTdZaDASqC, component: SubgraphInstanceManager
Nov 16 09:27:32.002 WARN no runtime host created, there is already a runtime host instantiated for this data source, address: 6261666b72656964643765377669706b36787978657078646a3468616268357a6e75717173326b65696e796b733433376f35326f69376536746561, name: MetadataInfo, sgd: 2737, subgraph_id: QmbrEG3Ph195yb341q1utKWu1WiupaeSeYRGRTdZaDASqC, component: SubgraphInstanceManager
thread 'QmbrEG3Ph195yb341q1utKWu1WiupaeSeYRGRTdZaDASqC[2737]' panicked at 'assertion failed: self.hosts.last().and_then(|h| h.creation_block_number()) <=
    data_source.creation_block()', /graph-node/core/src/subgraph/context/instance.rs:136:9
```

### IPFS hash

QmbrEG3Ph195yb341q1utKWu1WiupaeSeYRGRTdZaDASqC

### Subgraph name or link to explorer

_No response_

### Some information to help us out

- [ ] Tick this box if this bug is caused by a regression found in the latest release.
- [ ] Tick this box if this bug is specific to the hosted service.
- [X] I have searched the issue tracker to make sure this issue is not a duplicate.

### OS information

Linux

### PR DESCRIPTION
**Pull request by leoyvens**
#4810 was wrong. And that wasn't the only thing, the way the runner reverted in-memory state also didn't seem quite right. This adds a test which uncovered these issues. Should resolve #4994, unless that was hitting the assert for a different cause.

Needs an integration cluster run, of course.

### CODE REVIEW

File: core/src/subgraph/runner.rs [hunk 1]
```diff
@@ -1361,7 +1349,7 @@ where
             .deployment_head
             .set(subgraph_ptr.number as f64);
 
-        self.revert_state(subgraph_ptr.number)?;
+        self.revert_state_to(revert_to_ptr.number)?;
```
Comment (lutter) [inline core/src/subgraph/runner.rs:1352]: This is where reverting by more than one block came into play, right?

File: tests/src/fixture/mod.rs [hunk 2]
```diff
@@ -799,28 +798,47 @@ fn stream_events<C: Blockchain>(
...
+    struct ForkDb<B: Block> {
+        blocks: HashMap<BlockPtr, B>,
+    }
+
+    impl<B: Block> ForkDb<B> {
+        fn common_ancestor(&self, a: BlockPtr, b: BlockPtr) -> Option<&B> {
+            let mut a = self.blocks.get(&a).unwrap();
+            let mut b = self.blocks.get(&b).unwrap();
+            while a.number() > b.number() {
+                dbg!(a.ptr().number);
```
Comment (lutter) [inline tests/src/fixture/mod.rs:810]: I think that's a leftover `dbg!`

File: tests/src/fixture/mod.rs [hunk 3]
```diff
@@ -799,28 +798,47 @@ fn stream_events<C: Blockchain>(
...
+                a = self.blocks.get(&a.parent_ptr()?).unwrap();
+            }
+            while b.number() > a.number() {
+                dbg!(b.ptr().number);
```
Comment (lutter) [inline tests/src/fixture/mod.rs:814]: Same here

File: store/postgres/src/dynds/private.rs [hunk 4]
```diff
@@ -190,15 +190,14 @@ impl DataSourcesTable {
...
-        // This assumes all ranges are of the form [x, +inf).
+    pub(crate) fn revert(&self, conn: &PgConnection, block: BlockNumber) -> Result<(), StoreError> {
+        // Use the 'does not extend to the left of' operator `&>` to leverage the gist index, this
+        // is equivalent to lower(block_range) >= $1.
+        //
+        // This assumes all ranges are of the form [x, +inf), and thefore no range needs to be
+        // unclamped.
         let query = format!(
-            "delete from {} where block_range @> $1 and lower(block_range) >= $1",
+            "delete from {} where block_range &> int4range($1, null)",
```
[the remark this answers is not in the document]
Comment (leoyvens) [reply store/postgres/src/dynds/private.rs:200]: I don't think there is.

File: core/src/subgraph/runner.rs (hunk 1 above)
Comment (leoyvens) [reply core/src/subgraph/runner.rs:1352]: Yes this _and_ the implementation in `dynds/private.rs` were both wrong I think.

File: core/src/subgraph/runner.rs [hunk 5]
```diff
@@ -811,7 +803,6 @@ where
             // Handle unexpected stream errors by marking the subgraph as failed.
             Err(e) => {
                 self.metrics.stream.deployment_failed.set(1.0);
-                self.revert_state(block_ptr.block_number())?;
```
Comment (incrypto32) [inline core/src/subgraph/runner.rs:814]: Why was this revert_state removed?

**Pull request merged**